swift/Arrow/Sources/Arrow/ArrowCImporter.swift

```diff
@@ -0,0 +1,144 @@
+
+    public func importArray(
+        _ cArray: ArrowC.ArrowArray,
+        arrowType: ArrowType,
+        isNullable: Bool = false) -> Result<ArrowArrayHolder, ArrowError> {
+        let arrowField = ArrowField("", type: arrowType, isNullable: isNullable)
+        return importArray(cArray, arrowField: arrowField)
+    }
+
+    public func importArray(
+        _ cArray: ArrowC.ArrowArray,
+        arrowField: ArrowField) -> Result<ArrowArrayHolder, ArrowError> {
+        if cArray.null_count < 0 {
+            return .failure(.invalid("Uncomputed null count is not supported"))
+        } else if cArray.n_children > 0 {
+            return .failure(.invalid("Children currently not supported"))
+        } else if cArray.dictionary != nil {
+            return .failure(.invalid("Dictionary types currently not supported"))
+        } else if cArray.offset != 0 {
+            return .failure(.invalid("Offset of 0 is required but found offset: \(cArray.offset)"))
+        }
+
+        let arrowType = arrowField.type
+        let length = UInt(cArray.length)
+        let nullCount = UInt(cArray.null_count)
+        let nullBytes = Int(ceil(Double(length) / 8))
+        var arrowBuffers = [ArrowBuffer]()
+
+        if cArray.n_buffers > 0 {
+            if cArray.buffers == nil {
+                return .failure(.invalid("C array buffers is nil"))
+            }
+
+            switch arrowType.info {
+            case .variableInfo:
+                if cArray.n_buffers != 3 {
+                    return .failure(
+                        .invalid("Variable buffer count expected 3 but found \(cArray.n_buffers)"))
+                }
+
+                appendToBuffer(cArray.buffers[0], arrowBuffers: &arrowBuffers, byteCount: nullBytes, length: length,
+                               nullCount: nullCount)
+                let byteCount = MemoryLayout<Int32>.stride * Int(length + 1)
+                appendToBuffer(cArray.buffers[1], arrowBuffers: &arrowBuffers, byteCount: byteCount, length: length)
+                let offsetIndex = MemoryLayout<Int32>.stride * Int(length - 1)
+                let endIndex = arrowBuffers[1].rawPointer.advanced(by: offsetIndex).load(as: Int32.self)
+                appendToBuffer(cArray.buffers[2], arrowBuffers: &arrowBuffers, byteCount: Int(endIndex), length: length)
+            default:
+                if cArray.n_buffers != 2 {
+                    return .failure(.invalid("Expected buffer count 2 but found \(cArray.n_buffers)"))
+                }
+
+                appendToBuffer(cArray.buffers[0], arrowBuffers: &arrowBuffers, byteCount: nullBytes, length: length,
+                               nullCount: nullCount)
+                let byteCount = arrowType.getStride() * Int(length)
+                appendToBuffer(cArray.buffers[1], arrowBuffers: &arrowBuffers, byteCount: byteCount, length: length)
+            }
+        }
+
+        return makeArrayHolder(arrowField, buffers: arrowBuffers, nullCount: nullCount)
```

Comment:
Hmm, so when would the `release` pointer be called once the imported array is not referenced anymore?
It seems that currently you're leaking the imported array's memory...

Reply:
I may be misreading this code, bot how would the developer ensure the data is released when the array reference disappears?

Reply:
Maybe I am misunderstanding something.  The following is an example of the behavior I am expecting:
1. make array in go
2. make the carray in go
3. call ExportArrowArray to copy the go array to c array structure (this is still go)
4. call a method on the swift side that accepts the c array and imports the data into swift
5. the above method will perform some operation on the data or can cache the data for future reference.
6. the developers swift or go code would need to ensure that the memory is released once it is no longer needed
(on the swift side the ArrowCImporter has a method for releasing the schema and another for releasing the array)

Reply:
Hmm, so the general idea is that, once the producer has exported the data, the consumer is responsible for calling the `release` callback once it is done with the imported data.

When designing a higher-level API to abstract away details of the C Data Interface like this, the expectation is that the higher-level API level returns some kind of object (depending on the particular Arrow runtime) that ensure proper lifetime management of the data backing the imported array.

What "proper lifetime management" means depends on the language and runtime, but it should make it so that an imported array has similar semantics as a natively-allocated array. If a natively-allocated array has automatic memory management through reference counting, then so should an imported array.

For example, in Arrow C++, importing a C array backs the resulting C++ Array object with a dedicated [Buffer subclass](https://github.com/apache/arrow/blob/91bf1c9c170c1917ad47bb0dbb38aa5c9fbbbfb2/cpp/src/arrow/c/bridge.cc#L1462-L1481) that keeps the imported data alive and [calls the `release` pointer](https://github.com/apache/arrow/blob/91bf1c9c170c1917ad47bb0dbb38aa5c9fbbbfb2/cpp/src/arrow/c/bridge.cc#L1448-L1455) when the last `shared_ptr` reference to it vanishes. The result is that when an [imported C++ Array](https://github.com/apache/arrow/blob/91bf1c9c170c1917ad47bb0dbb38aa5c9fbbbfb2/cpp/src/arrow/c/bridge.h#L126-L128) is destroyed, the backing data automatically has its `release` pointer called. Just like, when a natively-allocated C++ Array is destroyed, the data backing its buffers is [automatically dellocated](https://github.com/apache/arrow/blob/91bf1c9c170c1917ad47bb0dbb38aa5c9fbbbfb2/cpp/src/arrow/memory_pool.cc#L844-L855).

From the looks of it, but I may be misreading, this PR currently doesn't ensure this equivalence between imported arrays and natively-allocated arrays.

Reply:
For example, if I were to write tests to exercise this using Python code, I would:
1. record the number of [PyArrow-allocated bytes](https://arrow.apache.org/docs/python/generated/pyarrow.total_allocated_bytes.html#pyarrow.total_allocated_bytes) as `orig_allocated`
1. allocate a PyArrow array
1. record the new number of PyArrow-allocated bytes as `new_allocated`
1. export the PyArrow array to a C array, and delete the Python array reference
1. import the C array into Swift
1. check that the number of PyArrow-allocated bytes is still `new_allocated`
1. delete the Swift array reference
1. check that the number of PyArrow-allocated bytes is back to `orig_allocated`

(you can imagine a similar scenario, but in reverse, to exercise importing into Python a Swift-exported array)

Reply:
Hi @pitrou,  gotcha, thank you for the detailed explanation!  I have updated the code to return a wrapped ArrayHolder that will manage releasing the memory associated with the C array.  I think this will resolve the issue.

Also, I altered the arrayToSwift go test to check if released was called to ensure memory was deallocated.  This is not as comprehensive as the test description above but it is a sanity check.  I will follow up with a future task to make this more robust as you have explained.  Let me know if this works.